website/public_adventures.py

```diff
@@ -31,67 +31,166 @@ def __init__(self, db: Database, achievements: Achievements):
-                "name": adventure.get("name"),
-                "short_name": adventure.get("name"),
-                "author": adventure.get("author", adventure["creator"]),
-                "creator": adventure.get("creator"),
-                "creator_public_profile": public_profile,
-                "date": utils.localized_date_format(adventure.get("date")),
-                "level": adventure.get("level"),
-                "levels": adventure.get("levels"),
-                "language": adv_lang,
-                "cloned_times": adventure.get("cloned_times"),
-                "tags": adv_tags,
-                "text": content,
-                "is_teacher_adventure": True,
-            }
-
-            # save adventures for later usage.
-            for _level in adventure.get("levels", [adventure.get("level")]):
-                _level = int(_level)
-                if self.adventures.get(_level):
-                    self.adventures[_level].append(current_adventure)
-                else:
-                    self.adventures[_level] = [current_adventure]
-
-            available_levels = adventure["levels"] if adventure.get("levels") else [adventure["level"]]
-            self.customizations["available_levels"].update([int(adv_level) for adv_level in available_levels])
-
     @route("/", methods=["GET"])
     @route("/filter", methods=["POST"])
     @requires_teacher
-    def filtering(self, user, index_page=False):
+    def filtering(self, user):
+        level = request.args["level"] if request.args.get("level") else "1"
+        language = request.args.get("lang")
+        if language == "reset":
+            language = ""
+        elif not language and request.method == 'GET':
+            language = g.lang
+        tag = request.args.get("tag", "")
+        search = request.form.get("search", request.args.get("search", ""))
+
+        available_languages = set()
+        available_levels = set()
+        available_tags = set()
+
+        # Get all possible filters
+        field_filters = self.db.get_public_adventures_filters()
+
+        available_levels.update(field_filters.get("level", []))
+        available_languages.update(field_filters.get("lang", []))
+        available_tags.update(field_filters.get("tag", []))
+
+        customizations = {"available_levels": sorted(available_levels, key=lambda x: int(x))}
+
+        tags = []
+        if tag:
+            toReset = request.args.get("reset")
+            if toReset:
+                # then it's the current selected tags, so remove current's tag
+                tags = [_tag for _tag in toReset.split(",") if _tag != tag]
+            else:
+                tags = tag.split(",")
+            tags = [t for t in tags if t]
+
+        # Get indexes
+        level_adventure_ids = self.db.get_public_adventures_indexes({"field_value": f"level_{level}"})
+        lang_adventure_ids = self.db.get_public_adventures_indexes({"field_value": f"lang_{language}"})
```

Comment:
It seems you are ignoring my suggestion to use a combined index for level and language, which would also allow pagination.

Can you put a short comment here for the people after you that you're not doing that, and the reasoning behind it? 

For example, a convincing motivation would look something like:

> Currently at XXX adventures, a search operation typically returns YYY MB. We will only start exceeding the 1MB download window at ZZZ adventures, which will probably take another AAA years. In the case of a combined (language, level) search, we won't be downloading more than BBBB bytes unnecessarily compared to a combined index, which adds at most CCC milliseconds. In comparison, the absence of having to write the combined index saves DDDD milliseconds for every adventure saved.

Reply:
It would be nicer to hide all the details of how to come up with this information in the `DATABASE` class, instead of putting it here. Pass the filter into a method in that class, and let it figure out how to come up with the information you need.